### ISSUE TITLE
Multi clients communication and some other questions.

### ISSUE THREAD

**Issue by skyhos** [warmcat/libwebsockets#243]
Hi, Thank you for the greate library.

I am a beginner of libwebsockets library and I tried to make a simple chat service via websocket on  Fedora Core 20 Linux server. I refered to [Martin Sikora's example](http://ahoj.io/libwebsockets-simple-websocket-server) for a client-server communication and I made it. Next, I was considering multi clients-server communications but I could not find any examples for it. I read [libwebsocket api document](https://libwebsockets.org/libwebsockets-api-doc.html) and studied the `libwebsockets.c`, `private-libwebsockets.h`, etc, and I figured out. My solution is replacing the `libwebsocket_write(...)` part in [Martin's example](http://ahoj.io/libwebsockets-simple-websocket-server) with following code :
```
  for(j=0;j<this->fds_count;j++){
      wsiTemp = this->lws_lookup[this->fds[j].fd]; // I felt awesome, indice are file decritors.
      libwebsocket_write(wsiTemp, &buf[LWS_SEND_BUFFER_PRE_PADDING], len, LWS_WRITE_TEXT);
    }
```
However, I do not think that this is the right solution. Because my solution needs the follwing code which was described in `private-libwebsockets.h` file to be specified or it print out the `dereferencing pointer to incomplete type` error while compiling via gcc.
```
#define LWS_MAX_SOCKET_IO_BUF 4096
struct libwebsocket_context {
	struct pollfd *fds;
	struct libwebsocket **lws_lookup; /* fd to wsi */
	int fds_count;
	int max_fds;
	int listen_port;
(...omitted...)
}
```
1. So, I ask to show me "standard" way if it exist.

2. Are there any documentations or books for the library? Although I can see the [libwebsocket api document](https://libwebsockets.org/libwebsockets-api-doc.html), I am hard to understand some concepts such as `wsi`, `user`, `fd`, etc in the document. For instance, I know that `wsi` means websocket instance, but I didn't know that `wsi` is used to distinguish clients. And I'd like to get more examples for the library.

3. While I build `libwebsockets-1.3` I met many `cmake` errors which require `zlib.h`, `in6addr.h`, `vfork.h` etc. Some are solved by intalling some packages such as `zlib-devel`, `openssl-devel`, `rpm-build` but others such as `vfork.h`, `in6addr.h` are not. So, I installed the lower version, `libwebsocket-1.22` by following [this installation guide](http://borjarefoyo.com/2013/07/06/install-libwebsockets-library-in-linux/). What packages does the `libwebsockets-1.3` depend on?

Thanks and have a good day.

**Comment by JoakimSoderberg** [comment]
1. Yes there is a standard way.

The example you're referring to using the dumb increment protocol is based on the [test-server](https://github.com/warmcat/libwebsockets/blob/master/test-server/test-server.c) that comes with libwebsockets, which already supports multiple clients.

You're using the API incorrectly, your program should not touch anything `private-libwebsockets.h`, hence the name `private`.

2. `wsi` is a per connection websocket instance. This is an opaque struct which you should not touch yourself except via the libwebsockets api.

3. If you're getting error please paste an error log here.... but to me it sounds more like the system introspection fails to find some headers, this is normal, these are just checks of what's available on the system. Unless it says ERROR and stops, there is no error.

**Comment by skyhos** [comment]
Thank you for the fast reply, Joakim.

I will check the test-server.c file.

I think I understood the meaning of ERROR incorrectly. ;-) I attach the logs : 
```
# cmake .. -DLIB_SUFFIX=64 -DLWS_IPV6=OFF

-- The C compiler identification is GNU 4.8.3
-- The CXX compiler identification is unknown
-- Check for working C compiler: /bin/cc
-- Check for working C compiler: /bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
CMake Error: your CXX compiler: "CMAKE_CXX_COMPILER-NOTFOUND" was not found.   Please set CMAKE_CXX_COMPILER to a valid compiler path or name.
-- CMAKE_TOOLCHAIN_FILE=''
-- Could NOT find Git (missing:  GIT_EXECUTABLE)
-- Performing Test HAVE_inline
-- Performing Test HAVE_inline - Success
-- Performing Test HAVE___inline__
-- Performing Test HAVE___inline__ - Success
-- Performing Test HAVE___inline
-- Performing Test HAVE___inline - Success
-- Looking for bzero
-- Looking for bzero - found
-- Looking for fork
-- Looking for fork - found
-- Looking for getenv
-- Looking for getenv - found
-- Looking for malloc
-- Looking for malloc - found
-- Looking for memset
-- Looking for memset - found
-- Looking for realloc
-- Looking for realloc - found
-- Looking for socket
-- Looking for socket - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for vfork
-- Looking for vfork - found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for dlfcn.h
-- Looking for dlfcn.h - found
-- Looking for fcntl.h
-- Looking for fcntl.h - found
-- Looking for in6addr.h
-- Looking for in6addr.h - not found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for netinet/in.h
-- Looking for netinet/in.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stdlib.h
-- Looking for stdlib.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for string.h
-- Looking for string.h - found
-- Looking for sys/prctl.h
-- Looking for sys/prctl.h - found
-- Looking for sys/socket.h
-- Looking for sys/socket.h - found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for vfork.h
-- Looking for vfork.h - not found
-- Looking for zlib.h
-- Looking for zlib.h - found
-- Looking for 4 include files stdlib.h, ..., float.h
-- Looking for 4 include files stdlib.h, ..., float.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of pid_t
-- Check size of pid_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.8")
ZLib include dirs: /usr/include
ZLib libraries: /usr/lib64/libz.so
Compiling with SSL support
-- Found OpenSSL: /usr/lib64/libssl.so;/usr/lib64/libcrypto.so (found version "1.0.1e")
OpenSSL include dir: /usr/include
OpenSSL libraries: /usr/lib64/libssl.so;/usr/lib64/libcrypto.so
Searching for OpenSSL executable and dlls
OpenSSL executable: /usr/bin/openssl
Generating SSL Certificates for the test-server...
Generating a 1024 bit RSA private key
..++++++
.................++++++
writing new private key to '/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/libwebsockets-test-server.key.pem'
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [XX]:State or Province Name (full name) []:Locality Name (eg, city) [Default City]:Organization Name (eg, company) [Default Company Ltd]:Organizational Unit Name (eg, section) []:Common Name (eg, your name or your server's hostname) []:Email Address []:SUCCSESFULLY generated SSL certificate
Generating API documentation
-- Looking for RPMTools... - found rpmuild is /bin/rpmbuild
-- RPMTools:: Using RPM_ROOTDIR=/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/RPM
-- Simple copy spec file <libwebsockets.spec> --> </root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/RPM/SPECS/libwebsockets.spec>
---------------------------------------------------------------------
  Settings:  (For more help do cmake -LH <srcpath>
---------------------------------------------------------------------
 LWS_WITH_SSL = ON  (SSL Support)
 LWS_SSL_CLIENT_USE_OS_CA_CERTS = 1
 LWS_USE_CYASSL = OFF (CyaSSL replacement for OpenSSL)
 LWS_WITHOUT_BUILTIN_GETIFADDRS = OFF
 LWS_WITHOUT_CLIENT = OFF
 LWS_WITHOUT_SERVER = OFF
 LWS_LINK_TESTAPPS_DYNAMIC = OFF
 LWS_WITHOUT_TESTAPPS = OFF
 LWS_WITHOUT_TEST_SERVER = OFF
 LWS_WITHOUT_TEST_SERVER_EXTPOLL = OFF
 LWS_WITHOUT_TEST_PING = OFF
 LWS_WITHOUT_TEST_CLIENT = OFF
 LWS_WITHOUT_TEST_FRAGGLE = OFF
 LWS_WITHOUT_DEBUG = OFF
 LWS_WITHOUT_EXTENSIONS = OFF
 LWS_WITH_LATENCY = OFF
 LWS_WITHOUT_DAEMONIZE = OFF
 LWS_USE_LIBEV =
 LWS_IPV6 = OFF
 LWS_WITH_HTTP2 = OFF
---------------------------------------------------------------------
-- Configuring incomplete, errors occurred!
See also "/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeOutput.log".
See also "/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeError.log".
```
they were something like [this mailing list](http://ml.libwebsockets.org/pipermail/libwebsockets/2014-May/001065.html) while running cmake. The solution of the mailing list does not work for me.

CMakeError.log :
```
Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler: CMAKE_CXX_COMPILER-NOTFOUND
Build flags:
Id flags:

The output was:
No such file or directory


Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler: CMAKE_CXX_COMPILER-NOTFOUND
Build flags:
Id flags: -c

The output was:
No such file or directory


Checking whether the CXX compiler is IAR using "" did not match "IAR .+ Compiler":
Determining if the include file in6addr.h exists failed with the following output:
Change Dir: /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp

Run Build Command:/bin/gmake "cmTryCompileExec3376686572/fast"
/bin/gmake -f CMakeFiles/cmTryCompileExec3376686572.dir/build.make CMakeFiles/cmTryCompileExec3376686572.dir/build
gmake[1]: Entering directory `/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp'
/usr/bin/cmake -E cmake_progress_report /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CMakeFiles 1
Building C object CMakeFiles/cmTryCompileExec3376686572.dir/CheckIncludeFile.c.o
/bin/cc  -DKEYWORD=__inline  -o CMakeFiles/cmTryCompileExec3376686572.dir/CheckIncludeFile.c.o   -c /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CheckIncludeFile.c
/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CheckIncludeFile.c:1:21: fatal error: in6addr.h: no such file or directory
 #include <in6addr.h>
                     ^
compilation terminated.
gmake[1]: *** [CMakeFiles/cmTryCompileExec3376686572.dir/CheckIncludeFile.c.o] ERROR 1
gmake[1]: Leaving directory `/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp'
gmake: *** [cmTryCompileExec3376686572/fast] ERROR 2


Determining if the include file vfork.h exists failed with the following output:
Change Dir: /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp

Run Build Command:/bin/gmake "cmTryCompileExec97611650/fast"
/bin/gmake -f CMakeFiles/cmTryCompileExec97611650.dir/build.make CMakeFiles/cmTryCompileExec97611650.dir/build
gmake[1]: Entering directory `/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp'
/usr/bin/cmake -E cmake_progress_report /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CMakeFiles 1
Building C object CMakeFiles/cmTryCompileExec97611650.dir/CheckIncludeFile.c.o
/bin/cc  -DKEYWORD=__inline  -o CMakeFiles/cmTryCompileExec97611650.dir/CheckIncludeFile.c.o   -c /root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CheckIncludeFile.c
/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp/CheckIncludeFile.c:1:19: fatal error: vfork.h: no such file or directory.
 #include <vfork.h>
                   ^
compilation terminated.
gmake[1]: *** [CMakeFiles/cmTryCompileExec97611650.dir/CheckIncludeFile.c.o] ERROR 1
gmake[1]: Leaving directory `/root/libwebsocketSource/libwebsockets-1.3-chrome37-firefox30/build/CMakeFiles/CMakeTmp'
gmake: *** [cmTryCompileExec97611650/fast] ERROR 2
```
Thanks, again.

**Comment by skyhos** [comment]
Well, I solve the problem myself : install `gcc-c++` package which is CXX compiler.

**Comment by 249708193o** [comment]
I have same problem.If someone knows how to resolve it, it would be great thanks for share.

**Comment by warmcat** [comment]
Sorry what problem do you have exactly?

You should be able to get around in6addr.h by disable ipv6 at cmake.

Quick hack for things like vfork is something like sudo touch /usr/include/vfork.h

**Comment by JoakimSoderberg** [comment]
Well the check for `vfork.h` is not an error. Just because it says that `vfork.h` not found, does not mean libwebsockets won't work, it simply means that the system introspection didn't find that.

I guess that check can be removed completely from the CMake project. The only reason it's there is because the old autoconf project had it. (And autoconf is usually configured by copying a start template that checks for a bunch of unnecessary stuff).

@warmcat: However, touching the file like that is dangerous on a system, because if you then compile some other software that looks for and actually uses `vfork`, it will think it's there and probably fail to compile.

@skyhos: As you figured out yourself the C++ compiler was the issue... However, I don't see why CMake even checks for that. Since the libwebsockets [CMakeLists.txt](CMakeLists.txt) contains this:

```c
project(libwebsockets C)
```

Which tells CMake this is a pure C project and it should not check for a C++ compiler:

> Optionally you can specify which languages your project supports. Example languages are CXX (i.e.  C++), C, Fortran, etc.  By default C and CXX are enabled.  E.g.  if you do not have a C++ compiler, you can disable the check for it by explicitely listing the languages you want to support, e.g.  C.  By using the special language "NONE" all checks for any language can be disabled.

**Comment by JoakimSoderberg** [comment]
And do you still get an **ERROR** because of `in6addr.h` in the console? Or do you mean in the `CMakeError.log` because that will contain all failed introspection tests.

Again, a system introspection test is not by default an ERROR, it is just a way to check what the system you are compiling on has available, and the code can be compiled based on that in mind.

**Comment by warmcat** [comment]
@JoakimSoderberg  yes.

**Comment by skyhos** [comment]
As I posted, my problem was solved by installing `gcc-c++`. However, I thought that warmcat made the library include `vfork.h` and `in6addr.h` because they are needed. When I was installing the library, I confused and thought that the problem was casued by lack of `vfork.h` and `in6aadr.h` since I removed the "not found `zlib.h`" message by installing `zlib-devel` package. To beginners such as me, it is hard to figure out the problem of the library, from installing to compiling successfully. It would be better if there are some documentation which describe some conceptions such as `wsi`, `rx_flow`, `user`, etc for beginners. I am still studying `test-server.c` with [the libwebsocket api document](https://libwebsockets.org/libwebsockets-api-doc.html), but I feel too hard.

Anyway, THANK YOU for the good library and your replies. : )
